casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala

```diff
@@ -0,0 +1,280 @@
+      incomingBlocksStream: Stream[F, BlockMessage],
+      finishedProcessingStream: Stream[F, BlockMessage],
+      casperShardConf: CasperShardConf
+  ): F[Stream[F, BlockHash]] = {
+
+    def blockStr(b: BlockMessage) = PrettyPrinter.buildString(b, short = true)
+    def logNotOfInterest(b: BlockMessage) =
+      Log[F].info(s"Block ${blockStr(b)} is not of interest. Dropped")
+    def logMalformed(b: BlockMessage) =
+      Log[F].info(s"Block ${blockStr(b)} is malformed. Dropped")
+//    def logMissingDeps(b: BlockMessage) = Log[F].info(s"Block ${blockStr(b)} missing dependencies.")
+
+    // CHeck if input string is equal to configuration shard ID
+    def checkIfEqualToConfigShardId(shardId: String) = {
+      val expectedShard = casperShardConf.shardName
+      val isValid       = expectedShard == shardId
+      def logMsg        = s"Ignored block with invalid shard, expected: $expectedShard, received: $shardId"
+      Log[F].info(logMsg).whenA(!isValid).as(isValid)
+    }
+
+    // Check if block data is cryptographically safe
+    def checkIntegrity(b: BlockMessage): F[Boolean] = Sync[F].defer {
+      val validShard = checkIfEqualToConfigShardId(b.shardId)
+      // TODO: 1. validation and logging in these checks should be separated
+      //       2. logging of these block should indicate that information cannot be trusted
+      //           e.g. if block hash is invalid it cannot represent identity of a block
+      val validFormat = Validate.formatOfFields(b)
+      val validHash   = Validate.blockHash(b)
+      val validSig    = Validate.blockSignature(b)
+      // TODO: check sender to be valid bonded validator
+      //  - not always possible because now are new blocks downloaded from DAG tips
+      //    which in case of epoch change sender can be unknown
+      validFormat &&^ validShard &&^ validHash &&^ validSig
+    }
+
+    // Check if block should be stored
+    def checkIfKnown(b: BlockMessage): F[Boolean] = {
+      val isStored = BlockStore[F].contains(b.blockHash)
+      val isTooOld = BlockDagStorage[F].getRepresentation.map { dag =>
+        dag.heightMap.headOption.map(_._1).forall(ProtoUtil.blockNumber(b) < _)
+      }
+      isStored ||^ isTooOld
+    }
+
+    def requestMissingDependencies(deps: Set[BlockHash]): F[Unit] =
+      deps.toList.traverse_(
+        BlockRetriever[F].admitHash(_, admitHashReason = BlockRetriever.MissingDependencyRequested)
+      )
+
+    // Process incoming blocks
+    def incomingBlocks(receiverOutputQueue: Queue[F, BlockHash]) =
+      incomingBlocksStream
+        .evalFilterAsyncUnorderedProcBounded { block =>
+          // Filter (ignore) blocks which doesn't pass integrity check
+          checkIntegrity(block).flatTap(logMalformed(block).unlessA(_))
+        }
+        .parEvalMapUnorderedProcBounded { block =>
+          // Start block checking, mark status in the state
+          val shouldCheck = state.modify(_.beginReceived(block.blockHash))
```

Comment:
Why `shouldCheck`? What should we check here? In `beginReceived` it is called `expectedReceive`. Need to check the blocks we expect to receive? Maybe rename it or add comment?

Reply:
I added more comments to clarify. This is synchronization to prevent multiple threads "working" on the same block. `shouldCheck` is a flag returned from the state which will only return true for one thread, all others should skip checking for the same block. It's like opening a transaction per block so subsequent threads should not do anything.